GuidedTours: add a design showcase tour

lsinger opened the pull request:
This PR adds a design showcase tour to show people around `/design`. 

TODO:

- [x] use `isAbTestInVariant` from #9819 once that's merged. 
- [x] rebase when #9817 is in `master` then use that shorthand for the ellipsis icon.

folletto commented:
Does this supersedes #7083?

lsinger commented:
> Does this supersedes #7083?

Yes, we'll take the best parts out of #7083 and then archive it.

lsinger commented:
@folletto this is the current state:

![theme-showcase-tour](https://cloud.githubusercontent.com/assets/23619/20570745/27ba4562-b1a5-11e6-933e-e38e8a488f5d.gif)

I.e. with the following steps:

- intro
- show paid / free switch
- search
- "more" button
- explain menu

Now I want to think about what steps we really need here, and what steps can go. As an expert on the design showcase — what features of it do you think are essential for users to become successful with it? To find a theme that they want? Devil's advocate: aren't "search" and "try & customize" all they need?

mcsf commented:
> what features […] are essential […] ? […] aren't "search" and "try & customize" all they need?

Search is being worked on (M5) and will be much more featureful for users — some of this can be [tried out in wpcalypso now](https://wpcalypso.wordpress.com/design) — and I suspect that will be a nice thing to showcase (pun unintended) in our tour. M5 is on hold as the team focuses on other efforts, but I'm sure @folletto can give us tips on estimates and/or what is good to show in a tour _now_.

folletto commented:
> Now I want to think about what steps we really need here, and what steps can go. As an expert on the design showcase — what features of it do you think are essential for users to become successful with it? To find a theme that they want? Devil's advocate: aren't "search" and "try & customize" all they need?

Yeah it seems the steps are the ones I made for the previous test, so I think it's a good starting point, but I agree I'd like something more compact too.

Iteration 3 of the showcase tutorial:

1. ( – baloon only )
**From this page you can change the design of your site. Want to see how to search for your ideal style?**
`[ Let's go! ]` `[ No thanks. ]`
2. ( ↓ arrow, box above )
**Here you can search the theme and apply filters.**
`[ Next ]` `[ Close ]`
3. ( ↑ arrow, box below )
**Here you can access all the extra theme options.**
`[ Next ]` `[ Close ]`
4. ( ↑ arrow, box below )
**This menu has everything you can do with a theme.**
`[ We're all done! ]`

In (2) I'm not sure if multi site has enough space to have the balloon above. If that doesn't fit, let's keep it below.

In (3) can we both trigger the next step on `[ Next ]` and on clicking •••?

These steps also should work well with M5, given the message is generic.

---

If I'm correct, the conditions triggering it are:

* New user AND
* Showcase Tour flag is Enabled AND
* Selected site is customizable AND (what is this about?)
* No in "customize".

To trigger it again, just append `?tour=designShowcase`, right?

I also noticed a bug:

* If I search for something else but "business" the next balloon doesn't show up (this shouldn't be an issue not even with the above that doesn't ask the user to type anymore).

lsinger commented:
> Iteration 3 of the showcase tutorial:

Thanks! Implemented your suggestions in 9c5290b. In the final step I added a sentence so as to have some sort of CTA at the end of the tour to give users an idea on what to try next. 

> In (2) I'm not sure if multi site has enough space to have the balloon above.

`above` is a bit buggy, so I'd rather not use it for now. But it's being worked on and should be ready soonish. /cc @marekhrabe 

> In (3) can we both trigger the next step on [ Next ] and on clicking •••?

We need the menu to open, so we need the user to click `•••`. 

> the conditions triggering it 

Close. :) It also had a bug, so this is the current state:

```JavaScript
isNewUser, // the user is a new user (note though that we set this to true for testing for now)
isEnabled( 'guided-tours/design-showcase' ), // the tour is enabled in the Calypso environment config file
isDesktop, // browser size is desktop size (width >960px)
selectedSiteIsCustomizable, // the current user can customize / change the theme for the selected site
not( inSection( 'customize' ) ), // we are not currently in the customizer
```

> To trigger it again, just append ?tour=designShowcase, right?

Right now just reloading `/design/<some wpcom site>` works for me. But yes, that's the you'd trigger it use the query argument. 

This is the current state of the tour: 

![theme-showcase-tour-v3](https://cloud.githubusercontent.com/assets/23619/20668940/b57de438-b571-11e6-83e1-5ac92dea76c4.gif)

@folletto what do you think? Does this introduce all the essential theme showcase elements and is generic enough to continue to work in the near future?

lsinger commented:
> Right now just reloading `/design/<some wpcom site>` works for me. 

Ha! That was my 2FA token that wasn't valid anymore, which I found out by navigating to `/me`. We save the preference "tours seen" to an endpoint that doesn't work if the 2FA token has expired.

folletto commented:
> In the final step I added a sentence so as to have some sort of CTA at the end of the tour to give users an idea on what to try next.

I'd lean in not doing that... because I'd like the user to actually try that menu. 
Does it make sense?


> We need the menu to open, so we need the user to click •••.

Ah yes. What I meant is: could we have also "Next", and trigger the menu open with that too? So programmatically opens it?
I'm thinking of this as it would mean having two steps both with next, and would create a nice resonance. 


> But yes, that's the you'd trigger it use the query argument.

Ok, awesome. It's very important it's always accessible manually via URL and not just triggering the criteria. For testing and support reasons.


> Does this introduce all the essential theme showcase elements and is generic enough to continue to work in the near future?

Yes. :)
And apart from the questions above, we are good. :)

lsinger commented:
> I'd lean in not doing that... because I'd like the user to actually try that menu.

I've changed the text of the additional sentence to "Try it out!" (referring to the menu) to bring it more in line with what you're trying to get at. Does that work for you? I believe we do need *some* sort of CTA here. 

> could we have also "Next", and trigger the menu open with that too?

That's not part of the framework so far, and I'd prefer to not add more framework features for now. Let's live with this version for now and possibly revisit in the future.

folletto commented:
> I've changed the text of the additional sentence to "Try it out!" (referring to the menu) to bring it more in line with what you're trying to get at. Does that work for you? I believe we do need some sort of CTA here.

Agreed, and I think you found the perfect solution. 👍  

> That's not part of the framework so far, and I'd prefer to not add more framework features for now. Let's live with this version for now and possibly revisit in the future.

Noted and agreed. Thanks. :)

folletto commented:
Design wise: 👍

folletto commented:
Testing on Chrome, there are a couple of... bugs?

(1) The ••• steps seems pointing to the wrong item when there are more than 2 columns:

![screen shot 2016-11-29 at 13 08 10](https://cloud.githubusercontent.com/assets/4389/20710942/e7879654-b634-11e6-89cb-08b2819e2f5e.png)

(2) If on that step I click on "Free" the balloon detaches...

![screen shot 2016-11-29 at 13 09 13](https://cloud.githubusercontent.com/assets/4389/20710971/0aff70d4-b635-11e6-9c89-f97e8d48e9b6.png)

lsinger commented:
So. Things / positions of things are pretty unpredictable. It all depends on the size of the browser window and on the scroll position. I tried to make the tour a bit more generic now to accomodate different situations. In practice, that means we have fewer arrow pointing (because we can't predict exactly where we'll need to point). 

@folletto can you give it another spin?

folletto commented:
Yeah the ••• is a bit... odd. Seems pointing to the second row in the test I did with 3 themes per row. If we detach it, would be possible to make it trigger for ANY ••• pressed, instead of the specific one?

Regardless, I think it's good to be merged.

One final aside question: how is`isNewUser` defined?

lsinger commented:
> would be possible to make it trigger for ANY ••• pressed

Not without changes to the framework, so I'd skip that for now. It's definitely something we should discuss as a future improvement as there are obviously cases — like this one — where it would make a lot of sense. 

> how is `isNewUser` defined?

Any user who registered a week or less ago:

```JavaScript
const WEEK_IN_MILLISECONDS = 7 * 1000 * 3600 * 24;
export const isNewUser = state => {
	const user = getCurrentUser( state );
	if ( ! user ) {
		return false;
	}

	const creation = Date.parse( user.date );
	return ( Date.now() - creation ) <= WEEK_IN_MILLISECONDS;
};
```

folletto commented:
Thanks! :shipit:

mcsf commented on client/layout/guided-tours/design-showcase-tour.js:
```diff
@@ -0,0 +1,104 @@
+} from 'state/ui/guided-tours/contexts';
+import Gridicon from 'components/gridicon';
+
+const isAbTestInVariant = function() {};
```
Prefer `lodash#noop`

lsinger replied:
Oops! This isn't actually supposed to go into this PR. I want to wait for #8627 to be merged so I can use `isAbTestInVariant` from there. Will add a TODO here to reflect that.

mcsf commented on client/layout/guided-tours/design-showcase-tour.js:
```diff
@@ -0,0 +1,104 @@
+			selectedSiteIsCustomizable,
+			not( inSection( 'customize' ) ),
+			isAbTestInVariant( 'designShowcaseTour', 'enabled' ),
+			) }
```
Minor: should we place the feature flag and A/B test checks first in the `and` list? For consistency.

lsinger replied:
I wanted the A/B test check to go *last* so it only ever gets called in case the user has the potential of getting the tour. If we make it go first we'll have many more users in the test, but many of those won't be getting the tour.

mcsf commented on client/layout/guided-tours/design-showcase-tour.js:
```diff
@@ -0,0 +1,104 @@
+			</p>
+			<ButtonRow>
+				<Next step="search">{ translate( "Let's go!" ) }</Next>
+				<Quit>{ translate( 'No thanks.' ) }</Quit>
```
Minor: copy: comma after "no"? Grammatically more correct, I think.

lsinger replied:
Not sure about that, actually. Since we're both non-native I suggest we'll wait for editorial's opinion here and leave as is for now.

mcsf commented on client/layout/guided-tours/design-showcase-tour.js:
```diff
@@ -0,0 +1,104 @@
+						components: {
+							GridIcon: <Gridicon icon="ellipsis" size={ 24 } />,
+						}
+					} )
```
See https://github.com/Automattic/wp-calypso/pull/9610/commits/dd31cd3cfaa51cc14063f944fad1adf38462dc36 to avoid inlining the icon stuff in the config:

```js
<Continue step="finish" target="…" icon="ellipsis" click />
```

lsinger replied:
I didn't get this to work. Looking at `Continue`'s `render`, I find this:

```JSX
return (
	<p className="guided-tours__actionstep-instructions">
		<em>{ this.props.children || translate( 'Click to continue.' ) }</em>
	</p>
);
```

I don't see how the `icon="..."` thing would work. Maybe we had that at *some* point?

lsinger replied:
Yup: #9817! 💥

mcsf commented on client/layout/guided-tours/design-showcase-tour.js:
```diff
@@ -0,0 +1,104 @@
+				{ 'Here you can search for themes and apply filters.' }
+			</p>
+			<ButtonRow>
+				<Next step="theme-options">{ translate( "Next" ) }</Next>
```
The explicit label isn't needed if we're just passing `"Next"`. Same for the `Quit` element.

lsinger replied:
The more you know! 😄👍

mcsf commented on client/layout/guided-tours/positioning.js:
```diff
@@ -156,8 +156,9 @@ function validatePlacement( placement, target ) {
-	const container = getScrollableSidebar();
+function scrollIntoView( target, scrollContainer ) {
+	// TODO(lsinger): consider replacing with http://yiminghe.me/dom-scroll-into-view/
+	const container = scrollContainer || getScrollableSidebar();
```
Does this work? In the config, I see `scrollContainer="body"`, but I don't see that string being turned into a DOM node.

lsinger replied:
Not here, that's right. But in `config-elements.js` we have:

```JavaScript
const stepPos = getStepPosition( { placement, targetSlug: target, shouldScrollTo, scrollContainer: this.scrollContainer } );
```

... and this is fine because:

```JavaScript
	componentWillMount() {
		...
		this.scrollContainer = query( this.props.scrollContainer )[ 0 ] || global.window;
		...
	}
```

mcsf commented:
>> would be possible to make it trigger for ANY ••• pressed

> Not without changes to the framework, so I'd skip that for now.

Actually, that should be possible thanks to https://github.com/Automattic/wp-calypso/pull/7304. Check your Redux debugger when toggling that ellipsis. :)

lsinger commented:
> Actually, that should be possible thanks to #7304.

Wheee! Nice. Taking a look.

lsinger commented:
>> Actually, that should be possible thanks to #7304.

> Wheee! Nice. Taking a look. 

See dd217a6.

lsinger commented on client/layout/guided-tours/design-showcase-tour.js:
```diff
@@ -0,0 +1,104 @@
+
+export const DesignShowcaseTour = makeTour(
+	<Tour
+		name="designShowcase"
```
Yup, that makes a lot of sense. Renamed the tour to `designShowcaseWelcome` in 0677df8.

mcsf commented on client/layout/guided-tours/design-showcase-welcome-tour.js:
```diff
@@ -0,0 +1,104 @@
+			not( inSection( 'customize' ) ),
+			isAbTestInVariant( 'designShowcaseWelcomeTour', 'enabled' ),
+			) }
+		>
```
For consistency with the `Step`s below, let's indent both lines here one increment to the left, `) }` and `>`.

mcsf commented on client/state/ui/guided-tours/contexts.js:
```diff
@@ -26,6 +27,9 @@ export const isNewUser = state => {
 
+export const userHasInteractedWithComponent = componentName => state =>
+	getLastAction( state ).component === componentName;
+
```
This is obviously subjective and debatable, but I think it's easier to achieve consistency if our naming style is more like `hasUserInteractedWithComponent`. The same idea would apply to the following selector in this file (`isSelectedSitePreviewable`).

It makes the return type immediately obvious (the func is a predicate), and would mean a more coherent global list of selectors in `state`. (/cc @aduth)

aduth replied:
@mcsf Agree. Had a similar dilemma with the naming of our capability checking selector, and opted to diverge from core function naming `current_user_can` for the reasons you mentioned, naming ours `canCurrentUser`.

https://github.com/Automattic/wp-calypso/blob/8247391/client/state/current-user/selectors.js#L89-L108
https://codex.wordpress.org/Function_Reference/current_user_can

mcsf commented on config/stage.json:
```diff
@@ -22,6 +22,7 @@
 		"community-translator": true,
 		"desktop-promo": true,
 		"guided-tours": true,
+		"guided-tours/design-showcase-welcome": true,
```
Is this one supposed to be true? Stage should aim to replicate production.

mcsf commented on config/wpcalypso.json:
```diff
@@ -26,6 +26,7 @@
 		"dev/test-helper": true,
 		"guided-tours": true,
 		"guided-tours/main": false,
+		"guided-tours/design-showcase-welcome": false,
```
This flag, along with the other instances in this PR where the value is false (i.e. all except for development.json) shouldn't be added.

mcsf commented:
🚢 🚢 🚢

lsinger commented:
Rebased and updated a/b test date and numbers, as well as the date in the tour version.
